hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenRejectHandler.java

```diff
@@ -126,9 +171,205 @@ public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckExcept
+        final var ledgerConfig = context.configuration().getConfigData(LedgerConfig.class);
+        final var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+        validateTrue(tokensConfig.tokenRejectEnabled(), NOT_SUPPORTED);
+        validateTrue(rejections.size() <= ledgerConfig.tokenRejectsMaxLen(), INVALID_TRANSACTION_BODY);
+
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var rejectingAccount = getIfUsableForAliasedId(
+                rejectingAccountID, accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+
+        // Process the rejections to prepare for the transfers and the removal of allowances
+        final var processedRejections =
+                processRejectionsForTransferAndAllowancesRemoval(rejections, context, rejectingAccount);
+        final var body = CryptoTransferTransactionBody.newBuilder()
+                .tokenTransfers(processedRejections.tokenTransferLists)
+                .build();
+
+        // Apply all changes to the handleContext's states by performing the transfer to the treasuries
+        final var transferContext = new TransferContextImpl(context, false, true);
+        final var steps = decomposeTransferIntoSteps(body, context.payer());
+        steps.forEach(step -> step.doIn(transferContext));
+
+        // Update the token allowances
+        if (hederaConfig.allowancesIsEnabled()) {
+            final var nftStore = context.writableStore(WritableNftStore.class);
+            updateAllowancesInState(rejectingAccount, processedRejections, nftStore, accountStore);
+        }
+    }
+
+    /**
+     * Processes the rejections specified in the transaction to prepare for the transfers and the removal of allowances.
+     *
+     * @param rejections The list of TokenReferences representing the rejections to process.
+     * @param context The HandleContext providing the current handling context.
+     * @param rejectingAccount The Account rejecting its tokens.
+     * @return ProcessedRejections containing the token transfer lists, updated fungible token allowances, and processed NFTs.
+     */
+    private ProcessedRejections processRejectionsForTransferAndAllowancesRemoval(
+            @NonNull final List<TokenReference> rejections,
+            @NonNull final HandleContext context,
+            @NonNull final Account rejectingAccount) {
+
+        // Initialize stores and get the rejecting account
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+        final var relStore = context.writableStore(WritableTokenRelationStore.class);
+        final var accountID = rejectingAccount.accountIdOrThrow();
+
+        // Prepare collections for processing
+        final var tokenTransferListMap = new HashMap<TokenID, TokenTransferList>();
```

Comment:
Got the feedback, looking at it again, I definitely see the problem with passing a mutable list and editing it, especially when we have multiple ones; it could be harder to read. So we may indeed utilize the Streams API.

I've refactored it and also excluded the logic for processing the allowances in this method, so now we directly return List<TokenTransferList> and just build the Transfers for the transactions.
I think it looks better now and is easier to read. Please let me know your feedback on the changes.